# Immune Repertoire Analysis Dashboard

Welcome to the Immune Repertoire Analysis Dashboard, a powerful and user-friendly R Shiny application designed to facilitate comprehensive analysis of antigen receptor CDR3 amino acid sequences. This dashboard enables users to explore their data through four main aspects: Data Input and Processing, Statistical Analysis, Network Analysis, and Phylogenetic Analysis. Below, you'll find an overview of each aspect, along with links to tutorial videos that demonstrate their functionalities.

## Table of Contents

- [Data Input and Processing](#data-input-and-processing)
- [Statistical Analysis](#statistical-analysis)
- [Network Analysis](#network-analysis)
- [Phylogenetic Analysis](#phylogenetic-analysis)
- [Tutorial Videos](#tutorial-videos)

## Data Input and Processing

The Data Input and Processing module allows users to upload their own files or integrate data from our database or their personal database. It includes:

- **File Upload**: Easily upload your immune repertoire data files.
- **Database Integration**: Seamlessly integrate your data with our extensive database or your own.
- **Anchor Clustering**: Run Anchor Clustering to group amino acid sequences (1 Hamming distance) and generate cluster parameters for network features. For more details, you can check Anchor Clustering github page: https://github.com/skylerchang/Anchor_Clustering_Nt.

## Statistical Analysis

The Statistical Analysis module provides various visualizations and analyses to understand the distribution and overlap of clonotypes:

- **Clonotype Distribution**: View clonotype distributions across different junction lengths, cell types, samples, and V and J genes.
- **Clonotype Overlap**: Check for clonotype overlaps across different samples, cell types, or tissues.

## Network Analysis

The Network Analysis module is the most engaging part of the dashboard. It allows users to:

- **Cluster Visualization**: Visualize all clusters with key network features.
- **Node Visualization**: Display nodes with different samples, cells, or V and J gene types.
- **Grouped Clusters**: Customize the visualization by specifying the level of spread and cluster size.
- **Interactive Clusters**: Click on any dot to explore the network connections of each cluster.
- **Interactive Visualization**: Show node-specific clonotypes within clusters.


## Phylogenetic Analysis

The Phylogenetic Analysis module provides insights into the evolutionary relationships of clonotypes:

- **Phylogenetic Relationships**: Explore the phylogenetic relationships of clonotypes within each cluster.
- **Multiple Sequence Alignment**: View multiple sequence alignment results to understand sequence similarities and differences.

## Tutorial Videos

For each module, we have prepared tutorial videos to guide you through their functionalities:

- &#128209; **Data Input and Processing Tutorial**
### Part1a:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/d7f4912f-58fa-4e2d-87a6-00bcae7f592c

### Part1b:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/65d8efc3-ad43-4ae1-8fa5-fe3072a31dd1

### Part1c:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/6afe0ba5-2b92-4da7-a9f6-5aa582f9d58d

- &#x1F4CA; **Statistical Analysis Tutorial**

### Part2:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/936aaaa3-a71f-4d77-a7e0-d6de2688e77a

- &#x1F517; **Network Analysis Tutorial**

### Part3a:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/cf9fcbea-421c-4d59-be7a-56e77c956b4f

### Part3b:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/7f682683-9348-48e2-986a-0f80ef0d382b

### Part3c:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/12342f58-27db-44c4-a59f-0d2a5fd4feea

- &#x1F332; **Phylogenetic Analysis Tutorial**

### Part4:
https://github.com/skylerchang/ClonotypeNetworkVis/assets/35077981/89a955b9-4e3d-4ed6-ac23-25a5c8aa0665

## Contact
We hope you find this dashboard useful for your immune repertoire analysis. If you have any questions or encounter any issues, please feel free to open an issue on this GitHub page. And feel free to contact with hchang02@uoguelph.ca.



